## Exploratory Data Analysis of Rossmann Pharmaceuticals Sales Data

### IMPORTING LIBRARIES

In [264]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

import sys,os
import warnings



In [265]:
sys.path.append(os.path.abspath(os.path.join('../scripts')))
from file_handler import FileHandler
from df_cleaner import *
from df_selector import *


In [266]:
!pip install plotly 


In [267]:
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', 50)

### DATA COLLECTION (Reading Data)

In [268]:
# create a FileHandler object
file_handler = FileHandler()

In [269]:
# reading the store csv file
store_df = file_handler.read_csv("../Data/store.csv")
store_df.head()

Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      2         a          a                570.0                       11.0   
2      3         a          a              14130.0                       12.0   
3      4         c          c                620.0                        9.0   
4      5         a          a              29910.0                        4.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              NaN              NaN   
1                    2007.0       1             13.0           2010.0   
2                    2006.0       1             14.0           2011.0   
3                    2009.0       0              NaN              NaN   
4                    2015.0       0              NaN              NaN   

     PromoInterval  
0              NaN  
1  Jan,Apr,Jul,Oct  
2  Jan,Apr,Jul,Oct  
3              NaN  
4              NaN

In [270]:
# reading the sales training csv file
train_df = file_handler.read_csv("../Data/train.csv")
train_df.head(10)

Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   
5      6          5  2015-07-31   5651        589     1      1            0   
6      7          5  2015-07-31  15344       1414     1      1            0   
7      8          5  2015-07-31   8492        833     1      1            0   
8      9          5  2015-07-31   8565        687     1      1            0   
9     10          5  2015-07-31   7185        681     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1  
5              1  
6              1  
7              1  
8              1  
9              1

In [271]:
# reading the sales training csv file
test_df = file_handler.read_csv("../Data/test.csv")
test_df.head(10)

Id  Store  DayOfWeek        Date  Open  Promo StateHoliday  SchoolHoliday
0   1      1          4  2015-09-17   1.0      1            0              0
1   2      3          4  2015-09-17   1.0      1            0              0
2   3      7          4  2015-09-17   1.0      1            0              0
3   4      8          4  2015-09-17   1.0      1            0              0
4   5      9          4  2015-09-17   1.0      1            0              0
5   6     10          4  2015-09-17   1.0      1            0              0
6   7     11          4  2015-09-17   1.0      1            0              0
7   8     12          4  2015-09-17   1.0      1            0              0
8   9     13          4  2015-09-17   1.0      1            0              0
9  10     14          4  2015-09-17   1.0      1            0              0

### DATA CLEANING

#### for Store Dataframe

In [165]:
store_df.shape

(1115, 10)

In [166]:
store_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1115 entries, 0 to 1114
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Store                      1115 non-null   int64  
 1   StoreType                  1115 non-null   object 
 2   Assortment                 1115 non-null   object 
 3   CompetitionDistance        1112 non-null   float64
 4   CompetitionOpenSinceMonth  761 non-null    float64
 5   CompetitionOpenSinceYear   761 non-null    float64
 6   Promo2                     1115 non-null   int64  
 7   Promo2SinceWeek            571 non-null    float64
 8   Promo2SinceYear            571 non-null    float64
 9   PromoInterval              571 non-null    object 
dtypes: float64(5), int64(2), object(3)
memory usage: 87.2+ KB


In [167]:
store_df.isnull().sum()

Store                          0
StoreType                      0
Assortment                     0
CompetitionDistance            3
CompetitionOpenSinceMonth    354
CompetitionOpenSinceYear     354
Promo2                         0
Promo2SinceWeek              544
Promo2SinceYear              544
PromoInterval                544
dtype: int64

In [168]:
store_df.describe()

Store  CompetitionDistance  CompetitionOpenSinceMonth  \
count  1115.00000          1112.000000                 761.000000   
mean    558.00000          5404.901079                   7.224704   
std     322.01708          7663.174720                   3.212348   
min       1.00000            20.000000                   1.000000   
25%     279.50000           717.500000                   4.000000   
50%     558.00000          2325.000000                   8.000000   
75%     836.50000          6882.500000                  10.000000   
max    1115.00000         75860.000000                  12.000000   

       CompetitionOpenSinceYear       Promo2  Promo2SinceWeek  Promo2SinceYear  
count                761.000000  1115.000000       571.000000       571.000000  
mean                2008.668857     0.512108        23.595447      2011.763573  
std                    6.195983     0.500078        14.141984         1.674935  
min                 1900.000000     0.000000         1.000000      2009.000000  
25%                 2006.000000     0.000000        13.000000      2011.000000  
50%                 2010.000000     1.000000        22.000000      2012.000000  
75%                 2013.000000     1.000000        37.000000      2013.000000  
max                 2015.000000     1.000000        50.000000      2015.000000

#### for train Data Frame 

In [169]:
train_df.shape

(1017209, 9)

In [170]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   Store          1017209 non-null  int64 
 1   DayOfWeek      1017209 non-null  int64 
 2   Date           1017209 non-null  object
 3   Sales          1017209 non-null  int64 
 4   Customers      1017209 non-null  int64 
 5   Open           1017209 non-null  int64 
 6   Promo          1017209 non-null  int64 
 7   StateHoliday   1017209 non-null  object
 8   SchoolHoliday  1017209 non-null  int64 
dtypes: int64(7), object(2)
memory usage: 69.8+ MB


In [171]:
train_df.isnull().sum()

Store            0
DayOfWeek        0
Date             0
Sales            0
Customers        0
Open             0
Promo            0
StateHoliday     0
SchoolHoliday    0
dtype: int64

In [172]:
train_df.describe()

Store     DayOfWeek         Sales     Customers          Open  \
count  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06   
mean   5.584297e+02  3.998341e+00  5.773819e+03  6.331459e+02  8.301067e-01   
std    3.219087e+02  1.997391e+00  3.849926e+03  4.644117e+02  3.755392e-01   
min    1.000000e+00  1.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    2.800000e+02  2.000000e+00  3.727000e+03  4.050000e+02  1.000000e+00   
50%    5.580000e+02  4.000000e+00  5.744000e+03  6.090000e+02  1.000000e+00   
75%    8.380000e+02  6.000000e+00  7.856000e+03  8.370000e+02  1.000000e+00   
max    1.115000e+03  7.000000e+00  4.155100e+04  7.388000e+03  1.000000e+00   

              Promo  SchoolHoliday  
count  1.017209e+06   1.017209e+06  
mean   3.815145e-01   1.786467e-01  
std    4.857586e-01   3.830564e-01  
min    0.000000e+00   0.000000e+00  
25%    0.000000e+00   0.000000e+00  
50%    0.000000e+00   0.000000e+00  
75%    1.000000e+00   0.000000e+00  
max    1.000000e+00   1.000000e+00


##### 1. HANDLING MISSING VALUES


##### for store dataframe


In [173]:
percent_missing_values(store_df)

The dataset contains 21.01 % missing values.


In [174]:
missing_df = missing_values_table(store_df)

Your selected dataframe has 10 columns.
There are 6 columns that have missing values.


In [175]:
missing_df

Missing Values  % of Total Values    Dtype
Promo2SinceWeek                       544              48.79  float64
Promo2SinceYear                       544              48.79  float64
PromoInterval                         544              48.79   object
CompetitionOpenSinceMonth             354              31.75  float64
CompetitionOpenSinceYear              354              31.75  float64
CompetitionDistance                     3               0.27  float64

From the missing table (missing_df), there is two groups with patterns. it means group with same number of missing values. first group is first 3 columns( Promo2sinceWeek,Promo2sinceYear,PromoInterval )and second group is next two columns(CompetitionOpenSinceMonth, CompetitionOpenSinceYear). 



In [176]:
#Number of rows with missing values for the whole dataset
count_missing_rows(store_df)

750 rows(67.26%) contain atleast one missing value.


In [177]:
# number of rows with missing values for group 1 (Promo2SinceWeek, Promo2SinceYear, and PromoInterval)
count_missing_rows(store_df[['Promo2SinceWeek', 'Promo2SinceYear', 'PromoInterval']])

544 rows(48.79%) contain atleast one missing value.


In [178]:
# number of rows with missing values for group 2 (CompetitionOpenSinceMonth, and CompetitionOpenSinceYear)
count_missing_rows(store_df[['CompetitionOpenSinceMonth', 'CompetitionOpenSinceYear']])

354 rows(31.75%) contain atleast one missing value.


As guessed we have got an interesting insight. Each columuns in these groups have missing values in the same rows as their groupmates. These means the missing values in these groups appear at the same time as a cluster. From the column names and the data fields' description in the challange document we can understand that the values of the columns in each group share a common information. Let's investigate.

In [185]:
# dataframe containing the missing rows for columns in group 1
group1_df = store_df[store_df['Promo2SinceWeek'].isna()]
group1_df.head(10)

Store StoreType Assortment  CompetitionDistance  \
0       1         c          a               1270.0   
3       4         c          c                620.0   
4       5         a          a              29910.0   
5       6         a          a                310.0   
6       7         a          c              24000.0   
7       8         a          a               7520.0   
8       9         a          c               2030.0   
9      10         a          a               3160.0   
15     16         a          c               3270.0   
22     23         d          a               4060.0   

    CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                         9.0                    2008.0       0   
3                         9.0                    2009.0       0   
4                         4.0                    2015.0       0   
5                        12.0                    2013.0       0   
6                         4.0                    2013.0       0   
7                        10.0                    2014.0       0   
8                         8.0                    2000.0       0   
9                         9.0                    2009.0       0   
15                        NaN                       NaN       0   
22                        8.0                    2005.0       0   

    Promo2SinceWeek  Promo2SinceYear PromoInterval  
0               NaN              NaN           NaN  
3               NaN              NaN           NaN  
4               NaN              NaN           NaN  
5               NaN              NaN           NaN  
6               NaN              NaN           NaN  
7               NaN              NaN           NaN  
8               NaN              NaN           NaN  
9               NaN              NaN           NaN  
15              NaN              NaN           NaN  
22              NaN              NaN           NaN

We can see that the Promo2 column is 0 for the first 10 rows. From intution we can infer that if there is no promotion the values for the columns in gropu 1(Promo2SinceWeek, Promo2SinceYear, and PromoInterval) should be null. Let's prove it.

In [180]:
# all unique values for each column
unique_values_df(group1_df)

Column  \
0                      Store   
1                  StoreType   
2                 Assortment   
3        CompetitionDistance   
4  CompetitionOpenSinceMonth   
5   CompetitionOpenSinceYear   
6                     Promo2   
7            Promo2SinceWeek   
8            Promo2SinceYear   
9              PromoInterval   

                                       Unique values  
0  [1, 758, 741, 742, 747, 751, 755, 757, 760, 4,...  
1                                       [a, d, c, b]  
2                                          [a, c, b]  
3  [250.0, 140.0, 50.0, 2640.0, 30.0, 420.0, 720....  
4  [9.0, 4.0, 11.0, 3.0, 12.0, 10.0, 7.0, 6.0, 2....  
5  [2013.0, 2012.0, 2014.0, 2009.0, 2010.0, 2005....  
6                                                [0]  
7                                                 []  
8                                                 []  
9                                                 []

In [181]:
group1_df['Promo2'].value_counts()

0    544
Name: Promo2, dtype: int64

All 544 values are 0. Now let's check the count of 0 in the Promo2 column for the whole store dataset.

In [182]:
store_df['Promo2'].value_counts()

1    571
0    544
Name: Promo2, dtype: int64

as we see, all rows with 0 value in their Promo2 columns
have null values in the columns Promo2SinceWeek, Promo2SinceYear, and PromoInterval.

Now let's impute the missing values in the columns Promo2SinceWeek and Promo2SinceYear with 0 to convey the meaning of absence since year and week can't be 0. For PromoInterval we should dig deeper what kind of data it holds since its data type is object


In [183]:
# check the exact data type of the object
pd.api.types.infer_dtype(store_df['PromoInterval'])

'string'

In [188]:
# all the unique values 
store_df['PromoInterval'].value_counts()

Jan,Apr,Jul,Oct     335
Feb,May,Aug,Nov     130
Mar,Jun,Sept,Dec    106
Name: PromoInterval, dtype: int64

For the column PromoInterval, we will impute it with '0,0,0,0' as the other values follow list-type format containing four months. So inorder to impute these columns we should first investigate if the values we will insert exist in each column as this will oppose our intention.

In [191]:
print ("for Promo2SinceWeek ",store_df[store_df['Promo2SinceWeek']== 0].shape)

print ("for Promo2SinceYear ",store_df[store_df['Promo2SinceYear'] == 0].shape)
print ("for Promo2SinceYear ",store_df[store_df['PromoInterval'] == '0,0,0,0'].shape)


for Promo2SinceWeek  (0, 10)
for Promo2SinceYear  (0, 10)
for Promo2SinceYear  (0, 10)


 *filling the missing values*

In [196]:
fix_missing_value(store_df, ['Promo2SinceWeek', 'Promo2SinceYear'], 0)

544 missing values in the column Promo2SinceWeek have been replaced by 0.
544 missing values in the column Promo2SinceYear have been replaced by 0.


In [197]:
fix_missing_value(store_df, ['PromoInterval'], '0,0,0,0')

544 missing values in the column PromoInterval have been replaced by 0,0,0,0.


In [198]:
# dataframe containing the missing rows for columns in group 1
group2_df = store_df[store_df['CompetitionOpenSinceMonth'].isna()]
group2_df.head(10)

Store StoreType Assortment  CompetitionDistance  \
11     12         a          c               1070.0   
12     13         d          a                310.0   
15     16         a          c               3270.0   
18     19         a          c               3240.0   
21     22         a          a               1040.0   
25     26         d          a               2300.0   
28     29         d          c               2170.0   
31     32         a          a               2910.0   
39     40         a          a                180.0   
40     41         d          c               1180.0   

    CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
11                        NaN                       NaN       1   
12                        NaN                       NaN       1   
15                        NaN                       NaN       0   
18                        NaN                       NaN       1   
21                        NaN                       NaN       1   
25                        NaN                       NaN       0   
28                        NaN                       NaN       0   
31                        NaN                       NaN       1   
39                        NaN                       NaN       1   
40                        NaN                       NaN       1   

    Promo2SinceWeek  Promo2SinceYear     PromoInterval  
11             13.0           2010.0   Jan,Apr,Jul,Oct  
12             45.0           2009.0   Feb,May,Aug,Nov  
15              0.0              0.0           0,0,0,0  
18             22.0           2011.0  Mar,Jun,Sept,Dec  
21             22.0           2012.0   Jan,Apr,Jul,Oct  
25              0.0              0.0           0,0,0,0  
28              0.0              0.0           0,0,0,0  
31             45.0           2009.0   Feb,May,Aug,Nov  
39             45.0           2009.0   Feb,May,Aug,Nov  
40             31.0           2013.0   Jan,Apr,Jul,Oct

In [199]:
# all unique values for each column in the dataframe
unique_values_df(group2_df)

Column  \
0                      Store   
1                  StoreType   
2                 Assortment   
3        CompetitionDistance   
4  CompetitionOpenSinceMonth   
5   CompetitionOpenSinceYear   
6                     Promo2   
7            Promo2SinceWeek   
8            Promo2SinceYear   
9              PromoInterval   

                                       Unique values  
0  [12, 853, 767, 766, 764, 762, 757, 756, 742, 7...  
1                                       [a, d, c, b]  
2                                          [c, a, b]  
3  [250.0, 2410.0, 50.0, 210.0, 840.0, 220.0, 150...  
4                                                 []  
5                                                 []  
6                                             [1, 0]  
7  [0.0, 40.0, 14.0, 45.0, 31.0, 22.0, 5.0, 1.0, ...  
8  [0.0, 2013.0, 2011.0, 2009.0, 2012.0, 2014.0, ...  
9  [0,0,0,0, Jan,Apr,Jul,Oct, Feb,May,Aug,Nov, Ma...

I can't see any pattern with the other columns. But from the column names and the data fields' description in the challange document we can understand that if CompetitionOpenSinceMonth and CompetitionOpenSinceYear are null it means there was already a competition when the store was opened. So we need impute the missing values in a meaningful way. Before doing that we will hold this meaning of an already existing competition in a separate column named CompetitionBeforeStoreOpened so that we can make use of this information in our analysis and prediction. This column will hold values of 0 and 1, 0 indicating the absence of competition at the time the store was opened and 1 indicating the presence of competition at the time the store was opened.

In [201]:
# deriving new column
store_df['CompetitionBeforeStoreOpened'] = store_df['CompetitionOpenSinceYear'].apply(lambda x: 1 if np.isnan(x) else 0)
store_df.head()

Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      2         a          a                570.0                       11.0   
2      3         a          a              14130.0                       12.0   
3      4         c          c                620.0                        9.0   
4      5         a          a              29910.0                        4.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              0.0              0.0   
1                    2007.0       1             13.0           2010.0   
2                    2006.0       1             14.0           2011.0   
3                    2009.0       0              0.0              0.0   
4                    2015.0       0              0.0              0.0   

     PromoInterval  CompetitionBeforeStoreOpened  
0          0,0,0,0                             0  
1  Jan,Apr,Jul,Oct                             0  
2  Jan,Apr,Jul,Oct                             0  
3          0,0,0,0                             0  
4          0,0,0,0                             0

Now we can impute the columns CompetitionOpenSinceMonth and CompetitionOpenSinceYear. So we will impute missing values in the column CompetitionOpenSinceYear and CompetitionOpenSinceMonth with its minimun value and 1 respectively. This kind of imputing assumes that the stores with missing values in these columns were opened after the first month of the minimum year value in the column CompetitionOpenSinceYear.

Let's find the minimum year value for CompetitionOpenSinceYear and impute the missing values.

In [202]:
min_year = store_df['CompetitionOpenSinceYear'].min()
min_year

1900.0

In [203]:
# impute the column CompetitionOpenSinceYear with 1900 and the column CompetitionOpenSinceMonth with 1
fix_missing_value(store_df, ['CompetitionOpenSinceYear'], min_year)
fix_missing_value(store_df, ['CompetitionOpenSinceMonth'], 1)

354 missing values in the column CompetitionOpenSinceYear have been replaced by 1900.0.
354 missing values in the column CompetitionOpenSinceMonth have been replaced by 1.


In [207]:
store_df.head(2)

Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      2         a          a                570.0                       11.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              0.0              0.0   
1                    2007.0       1             13.0           2010.0   

     PromoInterval  CompetitionBeforeStoreOpened  
0          0,0,0,0                             0  
1  Jan,Apr,Jul,Oct                             0

Now we are left with missing values in the column CompetitionDistance. Let's see the rows containing these missing values.

In [208]:
temp_df = store_df[store_df['CompetitionDistance'].isna()]
temp_df

Store StoreType Assortment  CompetitionDistance  \
290    291         d          a                  NaN   
621    622         a          c                  NaN   
878    879         d          a                  NaN   

     CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
290                        1.0                    1900.0       0   
621                        1.0                    1900.0       0   
878                        1.0                    1900.0       1   

     Promo2SinceWeek  Promo2SinceYear    PromoInterval  \
290              0.0              0.0          0,0,0,0   
621              0.0              0.0          0,0,0,0   
878              5.0           2013.0  Feb,May,Aug,Nov   

     CompetitionBeforeStoreOpened  
290                             1  
621                             1  
878                             1

In [209]:
unique_values_df(temp_df)

Column               Unique values
0                          Store             [291, 622, 879]
1                      StoreType                      [d, a]
2                     Assortment                      [a, c]
3            CompetitionDistance                          []
4      CompetitionOpenSinceMonth                       [1.0]
5       CompetitionOpenSinceYear                    [1900.0]
6                         Promo2                      [0, 1]
7                Promo2SinceWeek                  [0.0, 5.0]
8                Promo2SinceYear               [0.0, 2013.0]
9                  PromoInterval  [0,0,0,0, Feb,May,Aug,Nov]
10  CompetitionBeforeStoreOpened                         [1]

All rows with missing values in the column CompetitionDistance has a value of 1 in the column CompetitionBeforeStoreOpened. But this doesn't give us much information why the CompetitionDistance is null. But from common sense I guess this is because the competitions for these stores are very far away and have almost no impact on these stores that they weren't measured when collecting the data. For this reason, I will impute this values with the maximum competition distance.

In [210]:
max_dist = store_df['CompetitionDistance'].max()
max_dist

75860.0

In [211]:
fix_missing_value(store_df, ['CompetitionDistance'], max_dist)

3 missing values in the column CompetitionDistance have been replaced by 75860.0.


In [212]:
# final check for missing values
percent_missing_values(store_df)

The dataset contains 0.0 % missing values.


##### for Train dataframe 

In [214]:
percent_missing_values(train_df)

The dataset contains 0.0 % missing values.


We do not have any missing values in this dataset, we are lucky as it contains more than a million rows. 

#### 2. DATA TYPES 



checking if there is columns with mixed data types

##### Store dataframe

In [217]:
def show_cols_mixed_dtypes(df):
    mixed_dtypes = {'Column': [], 'Data type': []}
    for col in df.columns:
        dtype = pd.api.types.infer_dtype(df[col])
        if dtype.startswith("mixed"):
            mixed_dtypes['Column'].append(col)
            mixed_dtypes['Data type'].append(dtype)
    if len(mixed_dtypes['Column']) == 0:
        print('None of the columns contain mixed types.')
    else:
        print(pd.DataFrame(mixed_dtypes))

In [218]:
show_cols_mixed_dtypes(store_df)

None of the columns contain mixed types.


In [219]:
store_df.dtypes

Store                             int64
StoreType                        object
Assortment                       object
CompetitionDistance             float64
CompetitionOpenSinceMonth       float64
CompetitionOpenSinceYear        float64
Promo2                            int64
Promo2SinceWeek                 float64
Promo2SinceYear                 float64
PromoInterval                    object
CompetitionBeforeStoreOpened      int64
dtype: object

As we can see from the above data type of store the columns StoreType, Assortment and PromoInterval hold object. Even though PromoInterval holds list values, the are enclosed in string literals. We will convert the data types of these three columns to string. 

In [220]:
# get the columns with object data type
string_columns = store_df.select_dtypes(include='object').columns.tolist()
string_columns

['StoreType', 'Assortment', 'PromoInterval']

In [223]:
convert_to_string(store_df, string_columns)

the columns CompetitionOpenSinceMonth, CompetitionOpenSinceYear, Promo2SinceWeek, and Promo2SinceYear contain week, month and year information have float as dtype.  So we will change the data type of these columns to int64.

In [226]:
convert_to_int(store_df, ['CompetitionOpenSinceMonth',  'CompetitionOpenSinceYear',
        'Promo2SinceWeek', 'Promo2SinceYear'])

In [227]:
#Then we check again for dtypes of our dataframe
store_df.dtypes

Store                             int64
StoreType                        string
Assortment                       string
CompetitionDistance             float64
CompetitionOpenSinceMonth         int64
CompetitionOpenSinceYear          int64
Promo2                            int64
Promo2SinceWeek                   int64
Promo2SinceYear                   int64
PromoInterval                    string
CompetitionBeforeStoreOpened      int64
dtype: object

In [228]:
store_df.head()

Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                          9   
1      2         a          a                570.0                         11   
2      3         a          a              14130.0                         12   
3      4         c          c                620.0                          9   
4      5         a          a              29910.0                          4   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                      2008       0                0                0   
1                      2007       1               13             2010   
2                      2006       1               14             2011   
3                      2009       0                0                0   
4                      2015       0                0                0   

     PromoInterval  CompetitionBeforeStoreOpened  
0          0,0,0,0                             0  
1  Jan,Apr,Jul,Oct                             0  
2  Jan,Apr,Jul,Oct                             0  
3          0,0,0,0                             0  
4          0,0,0,0                             0

##### Train data frame 

In [231]:
#checking for the mixed data types
show_cols_mixed_dtypes(train_df)

         Column      Data type
0  StateHoliday  mixed-integer


as we see, the state Holiday container mixed interger, let check it unique values

In [232]:
train_df['StateHoliday'].value_counts()

0    855087
0    131072
a     20260
b      6690
c      4100
Name: StateHoliday, dtype: int64

In [233]:
train_df['StateHoliday'].value_counts().index

Index(['0', 0, 'a', 'b', 'c'], dtype='object')

 0 is represented as an integer and a string. Since there are other string values we will convert the StateHoliday column into string.

In [234]:
convert_to_string(train_df, ['StateHoliday'])

In [235]:
train_df.dtypes

Store             int64
DayOfWeek         int64
Date             object
Sales             int64
Customers         int64
Open              int64
Promo             int64
StateHoliday     string
SchoolHoliday     int64
dtype: object

In [236]:
train_df.head(2)

Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   

   SchoolHoliday  
0              1  
1              1

Date column contain datetime values but it has object dtype, so we have to convert it in datetime

In [239]:
convert_to_datetime(train_df, ['Date'])

In [240]:
train_df.dtypes

Store                     int64
DayOfWeek                 int64
Date             datetime64[ns]
Sales                     int64
Customers                 int64
Open                      int64
Promo                     int64
StateHoliday             string
SchoolHoliday             int64
dtype: object

#### 3. HANDLING DUPLICATES
##### For Store Dataframe

In [244]:
# search for duplicate rows and drop them
drop_duplicates(store_df)

No duplicate rows were found.


Since store is the id that uniquely identifies each record, there should be no duplicates. Let's check that.

In [245]:
store_df.duplicated(subset=['Store']).all()

False

We are safe. No two rows in our dataset contain the same Store value.

##### For Train Data Frame 

In [247]:
# search for duplicate rows and drop them
drop_duplicates(train_df)

No duplicate rows were found.


Store and Date are the columns that uniquely identify each record, there should be no duplicates. We can't have duplicate data for a store with the same date. Let's check whether such cases exist.

In [248]:
train_df.duplicated(subset=['Store', 'Date']).all()

False

We are safe. No two rows in our dataset contain the same combination of Store and Date values.

#### 4. FEATURE ENGINEERING 

##### Train dataframe 

Even though we have a date column it will be complex to query our data based on week, month, year, etc. So inorder tp gain complex insights easily, I have derived the following features.

In [256]:
train_df['Year'] = train_df['Date'].apply(lambda x: x.year)
train_df['Month'] = train_df['Date'].apply(lambda x: x.month)
train_df['DayOfMonth'] = train_df['Date'].apply(lambda x: x.day)
train_df['WeekOfYear'] = train_df['Date'].apply(lambda x: x.weekofyear)
train_df['weekday'] = train_df['DayOfWeek'].apply(lambda x: 0 if (x in [6, 7]) else 1)

In [257]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1017209 entries, 0 to 1017208
Data columns (total 14 columns):
 #   Column         Non-Null Count    Dtype         
---  ------         --------------    -----         
 0   Store          1017209 non-null  int64         
 1   DayOfWeek      1017209 non-null  int64         
 2   Date           1017209 non-null  datetime64[ns]
 3   Sales          1017209 non-null  int64         
 4   Customers      1017209 non-null  int64         
 5   Open           1017209 non-null  int64         
 6   Promo          1017209 non-null  int64         
 7   StateHoliday   1017209 non-null  string        
 8   SchoolHoliday  1017209 non-null  int64         
 9   Year           1017209 non-null  int64         
 10  Month          1017209 non-null  int64         
 11  DayOfMonth     1017209 non-null  int64         
 12  WeekOfYear     1017209 non-null  int64         
 13  weekday        1017209 non-null  int64         
dtypes: datetime64[ns](1), int64(12), s

In [259]:
train_df.sample(10)

Store  DayOfWeek       Date  Sales  Customers  Open  Promo  \
741611    922          4 2013-09-05   4637        524     1      0   
342008   1083          3 2014-09-10   4675        367     1      0   
161029    470          1 2015-03-09  10687       1486     1      0   
385454    486          5 2014-07-25   3290        428     1      0   
926496    717          6 2013-03-23   4266        445     1      0   
740607   1033          5 2013-09-06  11372       1212     1      0   
157248     34          4 2015-03-12   6675        882     1      0   
529532    693          5 2014-03-14   4347        385     1      0   
362141    563          2 2014-08-19   5994        767     1      1   
410022    488          7 2014-06-29      0          0     0      0   

       StateHoliday  SchoolHoliday  Year  Month  DayOfMonth  WeekOfYear  \
741611            0              0  2013      9           5          36   
342008            0              0  2014      9          10          37   
161029            0              0  2015      3           9          11   
385454            0              1  2014      7          25          30   
926496            0              0  2013      3          23          12   
740607            0              0  2013      9           6          36   
157248            0              0  2015      3          12          11   
529532            0              0  2014      3          14          11   
362141            0              1  2014      8          19          34   
410022            0              0  2014      6          29          26   

        weekday  
741611        1  
342008        1  
161029        1  
385454        1  
926496        0  
740607        1  
157248        1  
529532        1  
362141        1  
410022        0

In [261]:
def getMonth(month_list, index):
    month =  months = ['0', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sept', 'Oct', 'Nov', 'Dec']
    month_list = month_list.split(',')
    month = month_list[index]
    return months.index(month)

In [262]:
# split the PromoInterval column into 4 columns
store_df['PromoInterval0'] = store_df.PromoInterval.apply((lambda x: getMonth(x, 0)))
store_df['PromoInterval1'] = store_df.PromoInterval.apply((lambda x: getMonth(x, 1)))
store_df['PromoInterval2'] = store_df.PromoInterval.apply((lambda x: getMonth(x, 2)))
store_df['PromoInterval3'] = store_df.PromoInterval.apply((lambda x: getMonth(x, 3)))

In [263]:
store_df.sample(10)

Store StoreType Assortment  CompetitionDistance  \
536    537         a          a                600.0   
247    248         a          c                340.0   
165    166         a          c                100.0   
513    514         c          c               1200.0   
423    424         d          c               1250.0   
256    257         a          a                420.0   
409    410         c          a                 40.0   
264    265         a          a               4580.0   
226    227         a          a               2370.0   
4        5         a          a              29910.0   

     CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
536                          5                      2002       1   
247                          9                      2012       1   
165                          4                      2014       1   
513                          7                      2012       1   
423                          1                      1900       1   
256                         12                      2012       0   
409                         11                      2011       1   
264                          1                      1900       1   
226                          1                      1900       0   
4                            4                      2015       0   

     Promo2SinceWeek  Promo2SinceYear     PromoInterval  \
536                1             2012   Jan,Apr,Jul,Oct   
247               40             2012   Jan,Apr,Jul,Oct   
165               31             2013   Jan,Apr,Jul,Oct   
513               27             2012   Jan,Apr,Jul,Oct   
423               40             2011   Jan,Apr,Jul,Oct   
256                0                0           0,0,0,0   
409               22             2012  Mar,Jun,Sept,Dec   
264               14             2015   Jan,Apr,Jul,Oct   
226                0                0           0,0,0,0   
4                  0                0           0,0,0,0   

     CompetitionBeforeStoreOpened  PromoInterval0  PromoInterval1  \
536                             0               1               4   
247                             0               1               4   
165                             0               1               4   
513                             0               1               4   
423                             1               1               4   
256                             0               0               0   
409                             0               3               6   
264                             1               1               4   
226                             1               0               0   
4                               0               0               0   

     PromoInterval2  PromoInterval3  
536               7              10  
247               7              10  
165               7              10  
513               7              10  
423               7              10  
256               0               0  
409               9              12  
264               7              10  
226               0               0  
4                 0               0

#### Merging Dataframes And Saving Data

In [274]:
# save the clean dataframe to a csv file
file_handler.to_csv(train_df, '../Data/train.csv')
file_handler.to_csv(store_df, '../Data/store.csv')

In [275]:
# merging store_df and train_df
df = pd.merge(train_df, store_df, how='inner', on='Store')
df.shape

(1017209, 18)

In [276]:
# save merged dataframe
file_handler.to_csv(df, '../Data/train.csv')